File: packages/python/src/chromamark/inline.py
"""Inline ChromaMark constructs: [!pill], [.text], [=meter]."""

import re
from decimal import ROUND_HALF_UP, Decimal

from markdown_it.common.utils import escapeHtml

from .tones import parse_spec
from .whitespace import WS, WSR

SIGILS = {"!": "pill", ".": "text", "=": "meter"}

_SPLIT = re.compile(r"^([^" + WSR + r"]+)(?:[" + WSR + r"]+([\s\S]*))?$")
_PERCENT = re.compile(r"^([0-9]+(?:\.[0-9]+)?)[" + WSR + r"]*%$")
_FRACTION = re.compile(r"^([0-9]+(?:\.[0-9]+)?)[" + WSR + r"]*/[" + WSR + r"]*([0-9]+(?:\.[0-9]+)?)$")
_UNESCAPE = re.compile(r"\\([\s\S])")


def _unescape(text):
    return _UNESCAPE.sub(r"\1", text)


def _split_spec(inner):
    m = _SPLIT.match(inner)
    if not m:
        return None
    return m.group(1), (m.group(2) or "").strip(WS)


def _meter_width(value):
    percent = _PERCENT.match(value)
    fraction = _FRACTION.match(value)
    if percent:
        pct = float(percent.group(1))
    elif fraction:
        denom = float(fraction.group(2))
        if denom == 0:
            return None
        pct = float(fraction.group(1)) / denom * 100
    else:
        return None
    pct = max(0.0, min(100.0, pct))
    # Match JS String(+n.toFixed(2)): round the double half away from zero
    # (Decimal(pct) captures the exact double), then strip trailing zeros.
    quantized = Decimal(pct).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return format(quantized.normalize(), "f")


def _next_index(state, slot, needle, frm):
    """First index of `needle` at or after `frm`, memoized on the parse state so
    many openers on one long line stay O(line) overall instead of O(n^2)."""
    cache = getattr(state, slot, None)
    if cache is not None and cache[0] <= frm and (cache[1] == -1 or frm <= cache[1]):
        return cache[1]
    at = state.src.find(needle, frm)
    setattr(state, slot, (frm, at))
    return at


def _is_escaped(src, pos, floor):
    n = 0
    i = pos - 1
    while i >= floor and src[i] == "\\":
        n += 1
        i -= 1
    return n % 2 == 1


def _next_unescaped(state, slot, needle, frm, floor, max_):
    at = _next_index(state, slot, needle, frm)
    while at != -1 and at < max_ and _is_escaped(state.src, at, floor):
        at = _next_index(state, slot, needle, at + 1)
    return at


def _find_close(state, frm):
    max_ = state.posMax
    bracket = _next_unescaped(state, "_cm_br", "]", frm, frm, max_)
    if bracket == -1 or bracket >= max_:
        return -1
    newline = _next_unescaped(state, "_cm_nl", "\n", frm, frm, max_)
    if newline != -1 and newline < bracket:  # construct stays on one line
        return -1
    return bracket


def _make_rule(enabled):
    def chroma_inline(state, silent):
        start = state.pos
        src = state.src
        if start >= state.posMax or src[start] != "[":
            return False
        if start + 1 >= state.posMax:
            return False
        kind = SIGILS.get(src[start + 1])
        if not kind or not enabled.get(kind):
            return False

        end = _find_close(state, start + 2)
        if end == -1:
            return False

        inner = src[start + 2:end].strip(WS)
        parts = _split_spec(inner) if inner else None
        if not parts:
            return False
        spec_token, rest_raw = parts
        spec = parse_spec(spec_token)
        if not spec:
            return False
        rest = _unescape(rest_raw)

        if kind == "meter":
            if not rest:
                return False
            width = _meter_width(rest)
            if width is None:
                return False
            if not silent:
                tok = state.push("cm_meter", "", 0)
                tok.meta = {**spec, "value": rest, "width": width}
        elif kind == "text":
            if not rest:
                return False
            if not silent:
                tok = state.push("cm_text", "", 0)
                tok.meta = {**spec, "label": rest}
        else:
            label = rest or (spec_token.upper() if spec["tone"] else spec["color"])
            if not silent:
                tok = state.push("cm_pill", "", 0)
                tok.meta = {**spec, "label": label}

        state.pos = end + 1
        return True

    return chroma_inline


def _tone_attrs(meta):
    custom = " cm-custom" if meta.get("color") else ""
    tone = f' data-tone="{meta["tone"]}"' if meta.get("tone") else ""
    style = f' style="--fg:{escapeHtml(meta["color"])}"' if meta.get("color") else ""
    return custom, tone, style


def inline_plugin(md, enabled):
    md.inline.ruler.before("link", "cm_inline", _make_rule(enabled))

    def render_pill(self, tokens, idx, options, env):
        meta = tokens[idx].meta
        custom, tone, style = _tone_attrs(meta)
        return f'<span class="cm-pill{custom}"{tone}{style}>{escapeHtml(meta["label"])}</span>'

    def render_text(self, tokens, idx, options, env):
        meta = tokens[idx].meta
        custom, tone, style = _tone_attrs(meta)
        return f'<span class="cm-text{custom}"{tone}{style}>{escapeHtml(meta["label"])}</span>'

    def render_meter(self, tokens, idx, options, env):
        meta = tokens[idx].meta
        custom, tone, style = _tone_attrs(meta)
        aria = (
            ' role="progressbar" aria-valuemin="0" aria-valuemax="100"'
            f' aria-valuenow="{meta["width"]}" aria-valuetext="{escapeHtml(meta["value"])}"'
        )
        return (
            f'<span class="cm-meter{custom}"{tone}{style}{aria}>'
            f'<span class="cm-track"><span class="cm-fill" style="width:{meta["width"]}%"></span></span>'
            f'<span class="cm-val">{escapeHtml(meta["value"])}</span></span>'
        )

    md.add_render_rule("cm_pill", render_pill)
    md.add_render_rule("cm_text", render_text)
    md.add_render_rule("cm_meter", render_meter)

File: packages/renderer/src/inline.js
/**
 * Inline ChromaMark constructs sharing the `[<sigil>…]` family:
 *   [!spec label]  → pill / badge (filled)
 *   [.spec label]  → colored text (tint, no fill)
 *   [=spec value]  → progress meter (bar)
 *
 * The leading `spec` is a tone/alias, `color=<hex|name>`, or a bare `#hex`.
 * An unrecognized spec makes the rule decline so the text stays literal.
 */

import { parseSpec } from './tones.js';

const OPEN = 0x5b; // [
const SIGILS = { '!': 'pill', '.': 'text', '=': 'meter' };

function unescape(text) {
  return text.replace(/\\([\s\S])/g, '$1');
}

const WS = /\s/;

/**
 * Split an inline construct's body (state.src between `from` and `end`) into its
 * leading spec token and the start of the remaining content, without slicing the
 * whole — possibly huge — span first, so an invalid spec is rejected cheaply.
 * Equivalent to matching `^(\S+)(?:\s+…)?$` on the trimmed body.
 */
function splitInline(src, from, end) {
  let s = from;
  while (s < end && WS.test(src[s])) s++;
  let e = s;
  while (e < end && !WS.test(src[e])) e++;
  if (e === s) return null; // empty / all-whitespace body
  return { specToken: src.slice(s, e), restStart: e };
}

/** Compute a meter fill width (0–100, as a string) from a value like "87%" or "3/10". */
function meterWidth(value) {
  let pct;
  const percent = value.match(/^(\d+(?:\.\d+)?)\s*%$/);
  const fraction = value.match(/^(\d+(?:\.\d+)?)\s*\/\s*(\d+(?:\.\d+)?)$/);
  if (percent) {
    pct = parseFloat(percent[1]);
  } else if (fraction) {
    const denom = parseFloat(fraction[2]);
    if (denom === 0) return null;
    pct = (parseFloat(fraction[1]) / denom) * 100;
  } else {
    return null;
  }
  pct = Math.max(0, Math.min(100, pct));
  return String(+pct.toFixed(2));
}

/**
 * First index of `needle` (a single char) at or after `from`, memoized on the
 * parse state so N openers on one long line cost O(line) total rather than
 * O(N·line). The cache is valid because `state.src` is constant for a state and
 * the first occurrence at-or-after `from` is monotonic.
 */
function nextIndex(state, slot, needle, from) {
  const c = state[slot];
  if (c !== undefined && c.from <= from && (c.at === -1 || from <= c.at)) return c.at;
  const at = state.src.indexOf(needle, from);
  state[slot] = { from, at };
  return at;
}

/** True if `src[pos]` is escaped by an odd run of backslashes back to `floor`. */
function isEscaped(src, pos, floor) {
  let n = 0;
  for (let i = pos - 1; i >= floor && src.charCodeAt(i) === 0x5c; i--) n++;
  return (n & 1) === 1;
}

/** First unescaped `needle` at or after `from` (escapes counted from `floor`). */
function nextUnescaped(state, slot, needle, from, floor, max) {
  let at = nextIndex(state, slot, needle, from);
  while (at !== -1 && at < max && isEscaped(state.src, at, floor)) {
    at = nextIndex(state, slot, needle, at + 1);
  }
  return at;
}

function findClose(state, from) {
  const max = state.posMax;
  const bracket = nextUnescaped(state, '_cmBr', ']', from, from, max);
  if (bracket === -1 || bracket >= max) return -1;
  const newline = nextUnescaped(state, '_cmNl', '\n', from, from, max);
  if (newline !== -1 && newline < bracket) return -1; // stays on one line
  return bracket;
}

function makeRule(enabled) {
  return function chromaInline(state, silent) {
    const start = state.pos;
    if (state.src.charCodeAt(start) !== OPEN) return false;

    const kind = SIGILS[state.src[start + 1]];
    if (!kind || !enabled[kind]) return false;

    const end = findClose(state, start + 2);
    if (end === -1) return false;

    const src = state.src;
    const parts = splitInline(src, start + 2, end);
    if (!parts) return false;

    const spec = parseSpec(parts.specToken);
    if (!spec) return false;

    const rest = unescape(src.slice(parts.restStart, end).trim());

    let token;
    if (kind === 'meter') {
      if (!rest) return false;
      const width = meterWidth(rest);
      if (width === null) return false;
      if (!silent) {
        token = state.push('cm_meter', '', 0);
        token.meta = { ...spec, value: rest, width };
      }
    } else if (kind === 'text') {
      if (!rest) return false;
      if (!silent) {
        token = state.push('cm_text', '', 0);
        token.meta = { ...spec, label: rest };
      }
    } else {
      const label = rest || (spec.tone ? parts.specToken.toUpperCase() : spec.color);
      if (!silent) {
        token = state.push('cm_pill', '', 0);
        token.meta = { ...spec, label };
      }
    }

    state.pos = end + 1;
    return true;
  };
}

function toneAttrs(meta, escapeHtml) {
  const custom = meta.color ? ' cm-custom' : '';
  const tone = meta.tone ? ` data-tone="${meta.tone}"` : '';
  const style = meta.color ? ` style="--fg:${escapeHtml(meta.color)}"` : '';
  return { custom, tone, style };
}

export default function inlinePlugin(md, enabled) {
  md.inline.ruler.before('link', 'cm_inline', makeRule(enabled));
  const esc = md.utils.escapeHtml;

  md.renderer.rules.cm_pill = (tokens, idx) => {
    const { custom, tone, style } = toneAttrs(tokens[idx].meta, esc);
    return `<span class="cm-pill${custom}"${tone}${style}>${esc(tokens[idx].meta.label)}</span>`;
  };

  md.renderer.rules.cm_text = (tokens, idx) => {
    const { custom, tone, style } = toneAttrs(tokens[idx].meta, esc);
    return `<span class="cm-text${custom}"${tone}${style}>${esc(tokens[idx].meta.label)}</span>`;
  };

  md.renderer.rules.cm_meter = (tokens, idx) => {
    const meta = tokens[idx].meta;
    const { custom, tone, style } = toneAttrs(meta, esc);
    const aria =
      ` role="progressbar" aria-valuemin="0" aria-valuemax="100"` +
      ` aria-valuenow="${meta.width}" aria-valuetext="${esc(meta.value)}"`;
    return (
      `<span class="cm-meter${custom}"${tone}${style}${aria}>` +
      `<span class="cm-track"><span class="cm-fill" style="width:${meta.width}%"></span></span>` +
      `<span class="cm-val">${esc(meta.value)}</span>` +
      `</span>`
    );
  };
}
